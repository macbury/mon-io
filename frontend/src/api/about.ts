import { ApiBase } from './types'
import aboutQuery from './queries/about'
import {
  Query
} from './graphql'

export default class AboutApi extends ApiBase {
  public fetch = async () => {
    const { data } = await this.client.query<Query>({
      query: aboutQuery
    })

    return data.about
  }
}