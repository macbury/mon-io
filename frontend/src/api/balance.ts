import { ApiBase } from './types'

import {
  Query,
  QueryBalanceArgs,
  Aggregation
} from './graphql'

import getBalanceHistoryQuery from './queries/getBalanceHistory'
import getCurrentBalanceQuery from './queries/getCurrentBalance'

export default class BalanceApi extends ApiBase {
  public async getHistory(categoryId: string) {
    const { data: { balance } } = await this.client.query<Query, QueryBalanceArgs>({
      query: getBalanceHistoryQuery,
      variables: { categoryId, aggregation: Aggregation.Sum }
    })

    return balance
  }

  public async getHistoryPerMonth(categoryId: string) {
    const { data: { balance } } = await this.client.query<Query, QueryBalanceArgs>({
      query: getBalanceHistoryQuery,
      variables: { categoryId, aggregation: Aggregation.None }
    })

    return balance
  }

  public async getToday(categoryId: string) {
    const { data: { balance } } = await this.client.query<Query, QueryBalanceArgs>({
      query: getCurrentBalanceQuery,
      variables: { categoryId }
    })

    return balance?.today
  }
}
