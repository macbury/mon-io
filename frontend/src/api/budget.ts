import { ApiBase } from './types'
import { Moment } from 'moment-timezone'
import bind from 'bind-decorator'
import getBudgetQuery from './queries/getBudget'
import nextMonthQuery from './queries/summary/nextMonth'
import updateCategoryBudgetMutation from './mutations/updateCategoryBudget'

import {
  Query,
  Mutation,
  QueryGetBudgetArgs,
  UpdateCategoryBudgetInput,
  MutationUpdateCategoryBudgetArgs,
  QuerySummaryArgs
} from './graphql'

export default class BudgetApi extends ApiBase {
  public fetch = async (id : string) => {
    const { data } = await this.client.query<Query, QueryGetBudgetArgs>({
      query: getBudgetQuery,
      variables: { id }
    })

    return data.getBudget
  }

  @bind
  public async nextMonth(date : Moment) {
    const { data } = await this.client.query<Query, QuerySummaryArgs>({
      query: nextMonthQuery,
      variables: { date: date.format('YYYY-MM-01') }
    })

    return data.summary.nextMonth
  }

  @bind
  public async updateCategoryBudget(attributes : UpdateCategoryBudgetInput) {
    const { data: { updateCategoryBudget } } = await this.client.mutate<Mutation, MutationUpdateCategoryBudgetArgs>({
      mutation: updateCategoryBudgetMutation,
      variables: { input: attributes }
    })

    return updateCategoryBudget
  }
}