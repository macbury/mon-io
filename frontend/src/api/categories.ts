import bind from 'bind-decorator'
import { ApiBase } from './types'
import allCategoriesQuery from './queries/categories/all'
import updateCategoryMutation from './mutations/updateCategory'
import createCategoryMutation from './mutations/createCategory'

import {
  Query,
  Mutation,
  UpdateCategoryInput,
  MutationUpdateCategoryArgs,
  CreateCategoryInput,
  MutationCreateCategoryArgs
} from './graphql'

export default class CategoriesApi extends ApiBase {
  @bind
  public async getAllCategories() {
    const { data } = await this.client.query<Query>({
      query: allCategoriesQuery
    })

    return data.allCategories.nodes
  }

  @bind
  public async create(input : CreateCategoryInput) {
    try {
      const { data } = await this.client.mutate<Mutation, MutationCreateCategoryArgs>({
        mutation: createCategoryMutation,
        variables: {
          input
        }
      })

      return data.createCategory
    } catch (e) {
      return {
        category: null,
        errors: [e.toString()]
      }
    }
  }

  @bind
  public async update(input : UpdateCategoryInput) {
    try {
      const { data } = await this.client.mutate<Mutation, MutationUpdateCategoryArgs>({
        mutation: updateCategoryMutation,
        variables: {
          input
        }
      })

      return data.updateCategory
    } catch (e) {
      return {
        category: null,
        errors: [e.toString()]
      }
    }
  }
}