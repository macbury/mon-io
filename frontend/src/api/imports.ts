import createImportMutation from './mutations/createImport'
import importsQuery from './queries/imports'
import getImportQuery from './queries/getImport'
import updateImportMutation from './mutations/updateImport'
import destroyImportMutation from './mutations/destroyImport'

import { ApiBase } from './types'
import {
  Import,
  QueryGetImportArgs,
  QueryImportsArgs,
  CreateImportInput,
  UpdateImportInput,
  MutationCreateImportArgs,
  MutationDestroyImportArgs,
  MutationUpdateImportArgs,
  Mutation,
  Query,
  ImportConnection,
  CreateImportPayload,
  UpdateImportPayload,
  DestroyImportInput
} from './graphql'

export default class ImportsApi extends ApiBase {
  public async find(id: string, encoding?: string) : Promise<Import | null> {
    const { data: { getImport } } = await this.client.query<Query>({
      query: getImportQuery,
      variables: {
        id,
        encoding: encoding || ''
      }
    })

    return getImport
  }

  public async create(input : CreateImportInput) : Promise<CreateImportPayload> {
    try {
      const { data: { createImport } } = await this.client.mutate<Mutation, MutationCreateImportArgs>({
        mutation: createImportMutation,
        variables: { input }
      })

      return createImport
    } catch (error) {
      return {
        errors: [error.toString()],
        import: null
      }
    }
  }

  public async update(input : UpdateImportInput) : Promise<UpdateImportPayload> {
    try {
      const { data: { updateImport } } = await this.client.mutate<Mutation, MutationUpdateImportArgs>({
        mutation: updateImportMutation,
        variables: { input }
      })

      return updateImport
    } catch (error) {
      return {
        errors: [error.toString()],
        import: null
      }
    }
  }

  public async all(variables? : QueryImportsArgs) : Promise<ImportConnection> {
    const { data: { imports } } = await this.client.query<Query, QueryImportsArgs>({
      query: importsQuery,
      variables
    })

    return imports
  }

  public async destroy(input : DestroyImportInput) : Promise<boolean> {
    try {
      const { data: { destroyImport } } = await this.client.mutate<Mutation, MutationDestroyImportArgs>({
        mutation: destroyImportMutation,
        variables: { input }
      })

      return destroyImport.success
    } catch (error) {
      return false
    }
  }
}