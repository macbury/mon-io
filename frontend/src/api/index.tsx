import { ApolloClient } from 'apollo-client'
import { InMemoryCache } from 'apollo-cache-inmemory'
import { createHttpLink } from "apollo-link-http"
import { BatchHttpLink } from 'apollo-link-batch-http'
import { onError, ErrorResponse } from 'apollo-link-error'
import { ApolloLink } from 'apollo-link'
import { createUploadLink } from 'apollo-upload-client'

import promiseToObservable from '../helpers/promiseToObservable'
import ApolloRefreshToken from '../helpers/ApolloRefreshToken'
import logger from '../helpers/logger'
import RootStore from '../stores/RootStore'

import BalanceApi from './balance'
import CategoriesApi from './categories'
import ReceiptsApi from './receipts'
import SessionsApi from './session'
import TransactionsApi from './transactions'
import SummaryApi from './summary'
import AboutApi from './about'
import SeriesApi from './series'
import BudgetApi from './budget'
import ImportsApi from './imports'
import LocationApi from './location'

import hasFiles from '../helpers/hasFiles'

export default class MonioApi {
  private readonly log = logger('MonioApi')

  public readonly categories : CategoriesApi
  public readonly receipts : ReceiptsApi
  public readonly sessions : SessionsApi
  public readonly transactions : TransactionsApi
  public readonly summary : SummaryApi
  public readonly about : AboutApi
  public readonly budget : BudgetApi
  public readonly series : SeriesApi
  public readonly location : LocationApi
  public readonly imports : ImportsApi
  public readonly balance : BalanceApi

  private cache : InMemoryCache
  private store : RootStore
  private apolloRefreshToken : ApolloRefreshToken

  public client : ApolloClient<any>

  constructor(store : RootStore, client? : ApolloClient<any>) {
    this.store = store

    if (client) {
      this.client = client
    } else {
      this.setupClient()
    }

    this.categories = new CategoriesApi(this.client)
    this.receipts = new ReceiptsApi(this.client)
    this.sessions = new SessionsApi(this.client)
    this.transactions = new TransactionsApi(this.client)
    this.summary = new SummaryApi(this.client)
    this.about = new AboutApi(this.client)
    this.budget = new BudgetApi(this.client)
    this.series = new SeriesApi(this.client)
    this.location = new LocationApi(this.client)
    this.imports = new ImportsApi(this.client)
    this.balance = new BalanceApi(this.client)
  }

  private get linkOptions() {
    return {
      credentials: 'omit',
      fetch: this.apiFetch,
      fetchOptions: {
        method: 'POST',
        mode: 'cors'
      }
    }
  }

  private createHttpLink() {
    if (process.env.NODE_ENV == 'development') {
      return createHttpLink(this.linkOptions)
    } else {
      return new BatchHttpLink(this.linkOptions)
    }
  }

  private setupClient() {
    this.cache = new InMemoryCache()

    const uploadLink = createUploadLink(this.linkOptions)

    const httpLink = ApolloLink.split(
      ({ variables }) => hasFiles(variables),
      uploadLink,
      this.createHttpLink(),
    );

    this.apolloRefreshToken = new ApolloRefreshToken(this.store.session)

    const link = ApolloLink.from([
      this.onLoadingLink,
      this.onErrorLink,
      httpLink
    ])

    this.client = new ApolloClient({
      link,
      cache: this.cache,
      defaultOptions: {
        query: {
          fetchPolicy: 'no-cache',
        },
        watchQuery: {
          fetchPolicy: 'no-cache',
          errorPolicy: 'all'
        }
      }
    })
  }

  private apiFetch = async (_uri, options : RequestInit) => {
    // @ts-ignore
    options.headers.Authorization = 'Token token=' + this.store.session.accessToken
    // @ts-ignore
    options.headers.MonioClientVersion = this.store.about.clientVersion
    return fetch(this.store.session.endpointUrl, options)
  }

  private onLoadingLink = new ApolloLink((operation, forward) => {
    this.store.processing = true
    return forward(operation).map(response => {
      this.store.processing = false
      return response
    })
  })

  private onErrorLink = onError(({ graphQLErrors, networkError, forward, operation } : any) => {
    if (graphQLErrors)
      graphQLErrors.forEach(({ message, locations, path }) =>
        this.log(
          `[GraphQL error]: Message: ${message}, Location: ${locations}, Path: ${path}`,
        ),
      );

    if (networkError) {
      // @ts-ignore
      const { statusCode } = networkError

      if (statusCode === 400) {
        this.log(`[Outdated api error]: ${networkError}`)
        this.store.updateRequired = true
      } else if (statusCode === 401) {
        this.log("Refreshing access token!")
        return promiseToObservable(this.apolloRefreshToken.refreshToken()).flatMap((resp) => forward(operation))
      } else {
        this.log(`[Network error]: ${networkError}`, networkError?.result)
        this.store.processing = false
        this.store.setError(networkError?.result?.error)
      }
    }
  })
}