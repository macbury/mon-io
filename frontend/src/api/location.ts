import { ApiBase } from './types'
import getLocationQuery from './queries/getLocation'
import getMap from './queries/map'
import {
  Query,
  QueryGetLocationArgs,
  LocationInput,
  Location,
  LocationSummary,
  QueryMapArgs
} from './graphql'

export default class LocationApi extends ApiBase {
  public async reverseGeocode(location : LocationInput) : Promise<Location> {
    const { data } = await this.client.query<Query, QueryGetLocationArgs>({
      query: getLocationQuery,
      variables: { location }
    })

    return data.getLocation
  }

  public async getMap(categoryIds : string[]) : Promise<Array<LocationSummary>> {
    const { data } = await this.client.query<Query, QueryMapArgs>({
      query: getMap,
      variables: { categoryIds }
    })

    return data.map
  }
}