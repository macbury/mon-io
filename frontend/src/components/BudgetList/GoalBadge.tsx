import React from 'react'
import { Text } from 'react-native-paper'
import { useFormatMoney } from '../../helpers/currency'
import { Money, TransactionCategoryKind } from '../../api/graphql'
import styled from 'styled-components/native'
import Icon from '../Icon'
import { HeaderType, textColorByType } from './types'

export interface IGoalBadgeProps {
  amount: Money
  type: HeaderType
}

const InGoalBadge = styled(Text)`
  font-size: 18px;
  color: ${textColorByType};
  line-height: 27px;
`

const OutsideOfGoalBadge = styled.View`
  font-size: 16px;
  color: #fff;
  border-radius: 8px;
  padding: 4px 8px;
  line-height: 27px;
  background-color: ${textColorByType}aa;
  border-color: ${textColorByType};
  border-width: 1px;
  flex-direction: row;
`

const AttentionIcon = styled(Icon)`
  margin-right: 5px;
`

export default function GoalBadge({ amount, type } : IGoalBadgeProps) {
  const amountText = useFormatMoney(amount, true)

  if (amount.cents >= 0) {
    return (
      <InGoalBadge type={type}>
        {amountText}
      </InGoalBadge>
    )
  } else {
    const icon = type == TransactionCategoryKind.Income ? "MaterialCommunityIcons:check-circle" : "FontAwesome:exclamation-circle"
    return (
      <OutsideOfGoalBadge type={type}>
        <AttentionIcon
          size={16}
          name={icon}
          color="#fff" />
        <Text>{amountText}</Text>
      </OutsideOfGoalBadge>
    )
  }
}