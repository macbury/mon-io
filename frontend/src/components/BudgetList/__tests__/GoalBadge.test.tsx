import React from 'react'
import { shallow } from 'enzyme'
import snapshot from 'enzyme-to-json'
import { TransactionCategoryKind } from '../../../api/graphql'
import { money } from '../../../test/factories/money'

import GoalBadge, { IGoalBadgeProps } from '../GoalBadge'

it('renders expense with class', () => {
  const props : IGoalBadgeProps = {
    type: TransactionCategoryKind.Expense,
    amount: money.build({ cents: 890 })
  }

  const component = shallow(<GoalBadge {...props}/>)
  expect(snapshot(component)).toMatchSnapshot()
})

it('renders income with class', () => {
  const props : IGoalBadgeProps = {
    type: TransactionCategoryKind.Income,
    amount: money.build({ cents: 230 })
  }

  const component = shallow(<GoalBadge {...props}/>)
  expect(snapshot(component)).toMatchSnapshot()
})

it('renders below goal', () => {
  const props : IGoalBadgeProps = {
    type: TransactionCategoryKind.Income,
    amount: money.build({ cents: -230 })
  }

  const component = shallow(<GoalBadge {...props}/>)
  expect(snapshot(component)).toMatchSnapshot()
})