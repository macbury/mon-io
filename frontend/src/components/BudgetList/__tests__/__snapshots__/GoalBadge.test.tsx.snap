// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders below goal 1`] = `
<Styled(View)
  type="Income"
>
  <Styled(Icon)
    color="#fff"
    name="MaterialCommunityIcons:check-circle"
    size={16}
  />
  <withTheme(Text)>
    2.30 zł
  </withTheme(Text)>
</Styled(View)>
`;

exports[`renders expense with class 1`] = `
<Styled(withTheme(Text))
  type="Expense"
>
  8.90 zł
</Styled(withTheme(Text))>
`;

exports[`renders income with class 1`] = `
<Styled(withTheme(Text))
  type="Income"
>
  2.30 zł
</Styled(withTheme(Text))>
`;
